cmd/stolonctl/cmd/status.go

```diff
@@ -315,19 +338,50 @@ func generateStatus() (Status, error) {
 	}
 	status.Keepers = keepers
 
-	cluster := ClusterStatus{}
-	if cd.Cluster == nil || cd.DBs == nil {
-		cluster.Available = false
-	} else {
-		master := cd.Cluster.Status.Master
-		cluster.Available = true
+	return status, nil
+}
 
-		if master != "" {
-			cluster.MasterDBUID = cd.DBs[master].UID
-			cluster.MasterKeeperUID = cd.Keepers[cd.DBs[master].Spec.KeeperUID].UID
+func getKeeperRoles(dbuid string, cd *cluster.ClusterData) map[string]string {
+	var buildRoles func(parentDbuid, dbuid string, cd *cluster.ClusterData, keeperRoles map[string]string) map[string]string
+	buildRoles = func(parentDbuid, dbuid string, cd *cluster.ClusterData, keeperRoles map[string]string) map[string]string {
+		if dbuid == cd.Cluster.Status.Master {
+			keeperRoles[dbuid] = "master"
+		} else if isSynchronousStandby(parentDbuid, dbuid, cd) {
+			keeperRoles[dbuid] = "sync"
+		} else if isExternalSynchronousStandby(parentDbuid, dbuid, cd) {
+			keeperRoles[dbuid] = "external sync"
+		} else {
+			keeperRoles[dbuid] = "async"
+		}
+		db := cd.DBs[dbuid]
+		followers := db.Spec.Followers
```

Comment:
Instead of making a recursive function can you just iterate over all dbs and use their `followConfig` to detect the parent?